### PR TITLE
Stable merge for week 13 of 2021

### PR DESCRIPTION
**Pull request by raisjn**
new packages:
    [ddvk-hacks] Add ddvk-hacks (#247)
    [rmkit] add bufshot app, add lamp, add iago, add changelog (#276)
    [wikipedia] Initial wikipedia package.
    [deps] libdlib and libvncserver

updated packages:
    ~~[wireguard][1.0.20210219] - Updated package (and include wireguard-tools) (#285)~~
    [rm2fb] update rm2fb to work with xochitl 2.6 (v1.0.1) (#301)
    [recrossable] Update recrossable (#312)
    [appmarkable] Update appmarkable to 0.0.0-9 and rmservewacominput to 0.3.0-1 (#308) with rm2 support
    [rmkit] patch genie to fix crash in testing (#304)
    [oxide] Update Oxide to v2.1.2 (#241)
    [rm2fb] update rm2fb with wait ioctl and no-op on rM1 (#298)
    [rmkit] update rmkit to latest (2021-02-17) (#286)
    ~~[zshelf][0.3.1] - Updated Package (#287)~~

tooling:
    Pin the Ubuntu version used in workflows to 20.04 (#316)
    Provide better version number error messages (#314)
    util.auto_extract: Extract broken symlinks and missing directories (#302)
    build-time dependencies (#274).

### CODE REVIEW

**Review by erolx — DISMISSED** [review]

**Comment by Eeems** [comment]
@erolx did you approve this because you went through and tested everything? If so, what device did you test on, and what testing did you do?

**Comment by matteodelabre** [comment]
Some comments on the PR changelog:

* `wikipedia`, `bufshot`, `iago`, and `lamp` should be moved from the “updated packages” to the “new packages” list
* Packages `libdlib` and `libvncserver` are new.
* Another major tooling update that is included in this merge window is the addition of build-time dependencies (#274).

**Comment by matteodelabre** [comment]
I removed the `zoneinfo-utils` package along with the changes to the bootstrap script from this merge, for the same reasons as in the last merge (#288). This is blocked by #303.

**Comment by raisjn** [comment]
thank you for taking care of all of that! i cut this release one because i think our backlog is growing and it will get harder and harder to test / release. 

what command do you use for the git merge? (git merge testing -X theirs?) my logs tend to get weird with that merge and i thought it had too much commits, so i was trying to figure out the proper place to end

if @erolx did any testing, that would be great!

**Comment by matteodelabre** [comment]
> thank you for taking care of all of that! i cut this release one because i think our backlog is growing and it will get harder and harder to test / release.

I totally agree. Thanks for doing this!

> what command do you use for the git merge? (git merge testing -X theirs?) my logs tend to get weird with that merge and i thought it had too much commits, so i was trying to figure out the proper place to end

I usually branch off the latest stable commit and copy the changes from testing using `git checkout --no-overlay testing -- .`

File: package/evtest/package [hunk 1]
```diff
@@ -10,17 +10,14 @@ timestamp=2020-12-30T02:52Z
 section="utils"
 maintainer="Linus K. <linus@cosmos-ink.net>"
 license=GPL-2.0-only
+makedepends=(build:automake)
 
 image=base:v1.2.1
-source=(https://gitlab.freedesktop.org/libevdev/evtest/-/archive/evtest-1.34/evtest-evtest-1.34.zip)
+source=("https://gitlab.freedesktop.org/libevdev/evtest/-/archive/evtest-${pkgver%-*}/evtest-evtest-${pkgver%-*}.zip")
```
Comment (dcardin) [inline package/evtest/package:16]: This change to 1.34-2 breaks it for me.

Comment (Eeems) [reply package/evtest/package:16]: In what way? The build seems to run fine.

Comment (dcardin) [reply package/evtest/package:16]: Interesting. It should resolve to the same 1.34. 
I have checked out the branch on another system and will double check.

Comment (dcardin) [reply package/evtest/package:16]: I have no idea why that system was not building properly, but on the new one it builds properly. Thanks for checking with me though.

**Comment by raisjn** [comment]
tested on rM2 w. 2.5:

* ddvk hacks, install and uninstall - verify hacks show up in UI
* rmkit: bufshot, lamp, genie and iago: saved a fb.png with bufshot, verified it. called up iago using genie, then drew onto screen with lamp (a circle)
* wikipedia: installed and loaded wikipedia, opened keyboard. don't have a link to test blue on yet
* recrossable: installed, verified orderings of puzzle. drew letters into puzzle
* rmkit bump: mines (game over screen), harmony (grayscale colors), genie, remux 
* oxide: installed fret, decay, erode, oxide and tarnish, turned on oxide and launched / killed a bunch of apps, verified it works. didn't try any suspend stuff yet. did verify that lock screen works (already had a PIN setup)

did not test yet:

* libdlib, libvncserver
* wireguard
* appmarkable, rmservewacominput
* rm2fb

**Comment by raisjn** [comment]
when i uninstall oxide (& co), i don't get a message about how to turn xochitl back on. 

<details>

```
reMarkable: ~/.cache/toltec/ opkg remove oxide fret tarnish erode decay
Removing package oxide from root...
Reloading Oxide applications: Done!
Removing package fret from root...
Reloading Oxide applications: Done!
Removing package erode from root...
Reloading Oxide applications: Done!
Removing package decay from root...
Reloading Oxide applications: Done!
Collected errors:
 * print_dependents_warning: Package tarnish is depended upon by packages:
 * print_dependents_warning:    rot
 * print_dependents_warning:    decay
 * print_dependents_warning:    erode
 * print_dependents_warning: These might cease to work if package tarnish is removed.

 * print_dependents_warning: Force removal of this package with --force-depends.
 * print_dependents_warning: Force removal of this package and its dependents
 * print_dependents_warning: with --force-removal-of-dependent-packages.
reMarkable: ~/.cache/toltec/ opkg remove rot decade erode
Removing package rot from root...
```

</details>

**Comment by Eeems** [comment]
@raisjn Looks like you didn't remove tarnish

**Comment by raisjn** [comment]
thanks for catching that! it's a little bit of a mess how many packages there are to deal with - maybe a virtual package will help.

<details>

```
reMarkable: ~/.cache/toltec/ opkg install tarnish_2.1.2~1_armv7-3.2.ipk
Installing tarnish (2.1.2~1) to root...
Configuring tarnish.
reMarkable: ~/.cache/toltec/ systemctl enable tarnish --now
Created symlink /etc/systemd/system/multi-user.target.wants/tarnish.service → /lib/systemd/system/tarnish.service.
reMarkable: ~/.cache/toltec/ systemctl disable xochitl --now
reMarkable: ~/.cache/toltec/ opkg remove tarnish
Removing package tarnish from root...
Stopping tarnish
Disabling tarnish
Removed /etc/systemd/system/multi-user.target.wants/tarnish.service.
reMarkable: ~/.cache/toltec/
```

</details>

(EDIT: i don't see any useful messages in the above about turning xochitl back on)

**Comment by matteodelabre** [comment]
Testing on rM2 @ 2.5 too:

* **wireguard** — I tried to bring up a config using `wg-quick` and I got an error saying that `resolvconf` was not found. The system’s DNS seems to be managed by systemd, so we would need to have resolvectl installed and symlink resolvconf to resolvectl for this to work. Unfortunately, resolvectl is not available out of the box nor in Entware. @jonahweissman
* **rmservewacominput & appmarkable** — Tested the new gesture to quit _appmarkable_. Ran rmservewacominput and the pcWacomToMouse script on client side. Checked that I see output on the client when interacting with the pen.

I think we can consider that libdlib has been tested, since @raisjn tested recrossable and this app makes use of libdlib.

**Comment by jonahweissman** [comment]
>  **wireguard** — I tried to bring up a config using `wg-quick` and I got an error saying that `resolvconf` was not found. The system’s DNS seems to be managed by systemd, so we would need to have resolvectl installed and symlink resolvconf to resolvectl for this to work. Unfortunately, resolvectl is not available out of the box nor in Entware. @jonahweissman

@matteodelabre I am able to confirm this behavior. I can try writing a recipe for `resolvectl` sometime this week. (It seems like I might be able to adapt [the recipe from archlinux](https://github.com/archlinux/svntogit-packages/blob/packages/systemd/trunk/PKGBUILD).) In the meantime, would it better to remove wg-quick from the package or just include a message somewhere that the `interface.DNS` key is not currently supported?

**Comment by matteodelabre** [comment]
@jonahweissman Thanks for getting back to me.

> I am able to confirm this behavior. I can try writing a recipe for `resolvectl` sometime this week. (It seems like I might be able to adapt [the recipe from archlinux](https://github.com/archlinux/svntogit-packages/blob/packages/systemd/trunk/PKGBUILD).)

That would be very welcome. It’s a pity that we’ll have to build the whole of systemd just to get this single executable, though.

> In the meantime, would it better to remove wg-quick from the package or just include a message somewhere that the `interface.DNS` key is not currently supported?

This does not seem to be the only issue with `wg-quick`. I edited my config to remove the `DNS` key and I got the following error:

```console
$ wg-quick up wgconf
[#] ip link add wgconf type wireguard
[#] wg setconf wgconf /dev/fd/63
[#] ip link set mtu 1420 up dev wgconf
[#] wg set wgconf fwmark 51820
[#] ip -6 route add ::/0 dev wgconf table 51820
ip: either "to" is duplicate, or "table" is garbage
```

This config works fine on my computer, so I assume this is a difference with the `ip` command? Did anyone manage to make wg-quick work with a specific config?

**Comment by jonahweissman** [comment]
> This does not seem to be the only issue with `wg-quick`.
>
> This config works fine on my computer, so I assume this is a difference with the `ip` command? Did anyone manage to make wg-quick work with a specific config?

@matteodelabre I use a config like this:
```
[interface]
Address = 192.168.2.6/32
PrivateKey = ...

[peer]
PublicKey = ...
Endpoint = ...
AllowedIps = 192.168.2.0/24
```
and it's able to create the tunnel successfully:
```
systemd[1]: Starting WireGuard via wg-quick(8) for wg0...
wg-quick[1029]: [#] ip link add wg0 type wireguard
wg-quick[1029]: [#] wg setconf wg0 /dev/fd/63
wg-quick[1029]: [#] ip -4 address add 192.168.2.6/32 dev wg0
wg-quick[1029]: [#] ip link set mtu 1420 up dev wg0
wg-quick[1029]: [#] ip -4 route add 192.168.2.0/24 dev wg0
systemd[1]: Started WireGuard via wg-quick(8) for wg0.
```

Does it give you the same error when you start it with systemctl? It could also be a problem with using IPv6 addresses. In either case, this definitely needs to be fixed. I'm fine with not including wireguard in this stable merge to allow more time for debugging.

**Comment by matteodelabre** [comment]
@jonahweissman Can you try with `AllowedIPs = 0.0.0.0/0`? It seems that the error only occurs when using a wildcard IP.

**Comment by jonahweissman** [comment]
> Can you try with `AllowedIPs = 0.0.0.0/0`? It seems that the error only occurs when using a wildcard IP.

@matteodelabre Yep, I'm getting the same error now.

**Comment by matteodelabre** [comment]
I guess the difference is because of the following condition in `wg-quick`: <https://git.zx2c4.com/wireguard-tools/tree/src/wg-quick/linux.bash#n174>. When a wildcard IP is supplied, the script seems to use some tricks relying on policy-based routing to make everything go through the wg interface. I’m not sure why this is needed. In any case, the `ip` command bundled in BusyBox does not seem to be able to configure policy-based routing, which is what causes the error we’re seeing.

**Comment by matteodelabre** [comment]
I’m moving the WireGuard conversation to #323 and removing the changes to the `wireguard` package from this merge so that we can move forward with the rest of the packages.

**Comment by matteodelabre** [comment]
`wikipedia` and `ddvk-hacks` have been tested by @Eeems on rM1.

**Comment by matteodelabre** [comment]
* **rm2fb**
  * On rM1, checked that installing rm2fb works and that apps still work when prefixed with `rm2fb-client`.
  * On rM2 (system 2.5), installed KOReader v2021.01 (which is not included in this merge) and checked that the bug mentioned [in this comment](https://github.com/ddvk/remarkable2-framebuffer/issues/38#issue-795480847) is fixed by the new rm2fb version.
  * On rM2 (system 2.6), ran `rm2fb-client xochitl` and checked that the following log lines appear:

```
found update_fn at 311c00
found create_fn at 314168
found shutdown_fn at 31368c
found wait_fn at 31410c
```

**Comment by matteodelabre** [comment]
Testing on rM1:

* **rmkit**, **recrossable (& libdlib)**, **oxide**: followed the same testing protocol as @raisjn above.
* **rmservewacominput (& appmarkable)**: followed the same steps as on rM2 above.
* **libvncserver (via vnsee)**: connected to a VNC server through VNSee, checked that the display shows up properly and that I could interact with the touchscreen and the pen.

**Review by matteodelabre — APPROVED** [review]
I think we’re clear for merging.

**Review by Eeems — APPROVED** [review]
lgtm as well

**Pull request merged**